crates/sui-core/src/congestion_tracker.rs

```diff
@@ -0,0 +1,331 @@
+use moka::ops::compute::Op;
+use moka::sync::Cache;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use sui_types::base_types::ObjectID;
+use sui_types::effects::{InputSharedObject, TransactionEffects, TransactionEffectsAPI};
+use sui_types::execution_status::CongestedObjects;
+use sui_types::messages_checkpoint::{CheckpointTimestamp, VerifiedCheckpoint};
+use sui_types::transaction::{TransactionData, TransactionDataAPI};
+
+use crate::execution_cache::TransactionCacheRead;
+
+#[derive(Clone, Copy, Debug)]
+pub struct CongestionInfo {
+    pub last_cancellation_time: CheckpointTimestamp,
+
+    pub highest_cancelled_gas_price: u64,
+
+    pub last_success_time: Option<CheckpointTimestamp>,
+    pub lowest_executed_gas_price: Option<u64>,
+}
+
+impl CongestionInfo {
+    /// Update the congestion info with the latest congestion info from a new checkpoint
+    fn update_for_new_checkpoint(&mut self, new: &CongestionInfo) {
+        // If there are more recent cancellations, we need to know the latest highest
+        // cancelled price
+        if new.last_cancellation_time > self.last_cancellation_time {
+            self.last_cancellation_time = new.last_cancellation_time;
+            self.highest_cancelled_gas_price = new.highest_cancelled_gas_price;
+        }
+        // If there are more recent successful transactions, we need to know the latest lowest
+        // executed price
+        if new.last_success_time > self.last_success_time {
+            self.last_success_time = new.last_success_time;
+            self.lowest_executed_gas_price = new.lowest_executed_gas_price;
+        }
+    }
+
+    fn update_for_cancellation(&mut self, now: CheckpointTimestamp, gas_price: u64) {
+        self.last_cancellation_time = now;
```

Comment:
Does `now` ever differ from self.last_cancellation_time? Seems this is only called in the scope of one checkpoint.

Reply:
It can differ because there may not be any cancellations in the most recently processed checkpoint, in which case this wouldn't be called.